execution-engine/src/wasi/common.rs

```diff
@@ -373,6 +395,189 @@ pub struct WasiWrapper {
     exit_code: Option<u32>,
     /// Whether clock functions (`clock_getres()`, `clock_gettime()`) should be enabled.
     pub(crate) enable_clock: bool,
+    /// Whether strace is enabled.
+    pub(crate) enable_strace: bool,
+}
+
+// Convert a vector of bytes into a printable string.
+// TODO: Perhaps we should use a more generous definition of "printable".
+fn strace_string(bytes: &Vec<u8>, max: usize) -> String {
+    let mut res = String::from("\"");
+    let n = if bytes.len() > max { max } else { bytes.len() };
+    for i in 0..n {
+        if 0x20 <= bytes[i] && bytes[i] < 0x7f {
+            if bytes[i] == b'\\' || bytes[i] == b'"' {
+                res.push_str("\\");
+            }
+            res.push_str(&String::from_utf8_lossy(&bytes[i..i + 1].to_vec()))
+        } else if bytes[i] == 9 {
+            res.push_str("\\t")
+        } else if bytes[i] == 10 {
+            res.push_str("\\n")
+        } else if i + 1 < n && b'0' <= bytes[i + 1] && bytes[i + 1] <= b'9' {
+            // The following character is a digit, so use three octal digits.
+            res.push_str(&format!("\\{:03o}", bytes[i]))
+        } else {
+            res.push_str(&format!("\\{:o}", bytes[i]))
+        }
+    }
+    res.push_str("\"");
+    if bytes.len() > max {
+        res.push_str("...")
+    }
+    res
+}
+
+fn strace_iovec<T: MemoryHandler>(
+    success: bool,
+    memory_ref: &mut T,
+    address: u32,
+    base: u32,
+    count: u32,
+) {
+    if !success {
+        eprint!("_");
+        return;
+    }
+    if let Ok(len) = memory_ref.read_u32(address) {
+        let max = if len < 32 { len } else { 32 } as usize;
+        // This inefficiently copies everything, but it's only used for tracing.
+        if let Ok(bufs) = (|| {
+            let iovecs = memory_ref.unpack_array::<IoVec>(base, count)?;
+            memory_ref.read_iovec_scattered_in_reverse(&iovecs)
+        })() {
+            let mut buf: Vec<u8> = Vec::new();
+            for b in &bufs {
+                buf.extend(b)
+            }
+            eprint!("{}", strace_string(&buf, max))
+        } else {
+            eprint!("BAD_IOVEC")
+        }
+    } else {
+        eprint!("BAD_IOVEC_LEN") // This will probably never happen.
+    }
+}
+
+macro_rules! strace_arg {
+    ($success: expr, ($mem: expr, $adr: expr, $len: expr, buffer)) => {
+        match $mem.read_buffer($adr, $len) {
+            Ok(x) => eprint!("{}", strace_string(&x, 32)),
+            Err(_) => eprint!("BAD_MEM_REF"),
+        }
+    };
+    ($success: expr, ($n: expr, dec)) => {
+        eprint!("{}", $n)
```

Comment:
I'm wary of us having `eprintln` and friends directly in the runtime itself.  I think we've taken pains to avoid that until now whenever possible.  Rather than printing directly to `stderr` is there no way of printing to a dedicated `strace` stream which can be read after the program has finished executing, like `stdout` and `stderr` are, with access managed by the policy?

Reply:
But what if the runtime panics, or the program never finishes? It would be nice to get strace output when that happens.

Reply:
Hmm, this is a good point that I didn't consider.  I think, without introducing some complex scaffolding, that you're right that `eprintln!` is the best that can be done for tracing non-terminating or crashing programs.